servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,372 @@
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            private final ByteToMessageDecoder decoder = decoderSupplier.get();
+            private final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+            @Nullable
+            PublisherSource.Subscription subscription;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                this.subscription = subscription;
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream decoder previously closed but more input arrived");
+                }
+
+                if (src == null) {
+                    throw new NullPointerException("Cannot decode null values");
+                }
+
+                try {
+                    // onNext will produce AT-MOST N items (as received)
+                    channel.writeInbound(toByteBuf(src));
+                    Buffer buffer = drainChannelQueueToSingleBuffer(channel.inboundMessages(), allocator);
+
+                    if (buffer != null && buffer.readableBytes() > 0) {
+                        subscriber.onNext(buffer);
+                    } else {
+                        // Not enough data to decompress, ask for more
+                        subscription.request(1);
+                    }
+                } catch (Throwable e) {
+                    throw new CodecDecodingException(NettyChannelContentCodec.this, e);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                safeCleanup(channel);
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                try {
+                    cleanup(channel);
+                } catch (Throwable t) {
+                    subscriber.onError(new CodecDecodingException(NettyChannelContentCodec.this, t));
+                    return;
+                }
+
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private EmbeddedChannel newEmbeddedChannel(final ChannelHandler handler, final BufferAllocator allocator) {
+        final EmbeddedChannel channel = new EmbeddedChannel(handler);
+        channel.config().setAllocator(getByteBufAllocator(allocator));
+        return channel;
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer) {
+        return BufferUtils.extractByteBufOrCreate(buffer);
+    }
+
+    private static ByteBuf toByteBuf(final Buffer buffer, final int offset, final int length) {
+        final Buffer slice = buffer.slice(buffer.readerIndex() + offset, length);
+        return BufferUtils.extractByteBufOrCreate(slice);
```

Comment:
If you take the slice instead of doing checks by yourself, these methods are not necessary and `extractByteBufOrCreate` can be used.